### ISSUE TITLE
compute_mask not called when custom layer does not alter inputs and mask is not None

### ISSUE THREAD

**Issue by deasmhumhna** [tensorflow/tensorflow#32322]
<em>Please make sure that this is a bug. As per our [GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md), we only address code/doc bugs, performance issues, feature requests and build/installation issues on GitHub. tag:bug_template</em>

**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04):
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device:
- TensorFlow installed from (source or binary): binary
- TensorFlow version (use command below): 2.0.0-beta
- Python version: 3.7.3-2
- Bazel version (if compiling from source):
- GCC/Compiler version (if compiling from source):
- CUDA/cuDNN version:
- GPU model and memory:

You can collect some of this information using our environment capture
[script](https://github.com/tensorflow/tensorflow/tree/master/tools/tf_env_collect.sh)
You can also obtain the TensorFlow version with: 1. TF 1.0: `python -c "import
tensorflow as tf; print(tf.GIT_VERSION, tf.VERSION)"` 2. TF 2.0: `python -c
"import tensorflow as tf; print(tf.version.GIT_VERSION, tf.version.VERSION)"`

**Describe the current behavior**
When the layer does not alter the inputs (but alters the mask), if `mask` is not `None`, the layer does not execute the `compute_mask` method. 
**Describe the expected behavior**
Call `compute_mask` and attach the mask to output regardless.
**Code to reproduce the issue**
Provide a reproducible test case that is the bare minimum necessary to generate the problem.
```
import tensorflow as tf
import numpy as np

class RandomMaskingNoAlter(tf.keras.layers.Layer):

  def call(self, inputs):
    return inputs

  def compute_mask(self, inputs, mask=None):
    print('no alter executed')
    if mask is None:
      return None
    random_mask = tf.cast(tf.random.uniform(tf.shape(mask), 0, 1, dtype=tf.int32), tf.bool)
    return tf.math.logical_and(random_mask, mask)

class RandomMaskingAlter(tf.keras.layers.Layer):
    
  def call(self, inputs):
    return inputs + 0

  def compute_mask(self, inputs, mask=None):
    print('alter executed')
    if mask is None:
      return None
    random_mask = tf.cast(tf.random.uniform(tf.shape(mask), 0, 1, dtype=tf.int32), tf.bool)
    return tf.math.logical_and(random_mask, mask)

x = np.array([[1, 4, 2, 2, 0, 0], [1, 1, 1, 0, 0, 0], [3, 2, 2, 3, 4, 1]], dtype='i4')
y = tf.keras.layers.Embedding(5, 5, mask_zero=True)(x)

z1 = RandomMaskingNoAlter()(x)
z2 = RandomMaskingNoAlter()(y)
z3 = RandomMaskingAlter()(x)
z4 = RandomMaskingAlter()(y)
```
**Other info / logs**
Include any logs or source code that would be helpful to diagnose the problem. If including tracebacks, please include the full traceback. Large logs and files should be attached.

**Comment by oanush** [comment]
@deasmhumhna ,
When tried executing the given code the following output was received 
`no alter executed
 alter executed
 alter executed` is the same output received from your side also?Thanks!

**Comment by deasmhumhna** [comment]
> @deasmhumhna ,
> When tried executing the given code the following output was received
> `no alter executed alter executed alter executed` is the same output received from your side also?Thanks!

Yes, that is what I get. The result should be `no alter executed no alter executed alter executed alter executed` (`compute_mask` should always be called if implemented by a subclass). I am however curious why it runs when `mask=None` but not if there is a pre-existing mask. Some kind of optimization?

**Comment by oanush** [comment]
Issue replicating with TF 2.0beta, kindly find the [gist](https://colab.sandbox.google.com/gist/oanush/bbbb2f8fd4f731bb259b71f4370a2a4a/32322.ipynb) of colab.Thanks!

**Comment by ravikyram** [comment]
Issue is replicating with TF nightly version (`2.4.0-dev20200902`).Please, find the gist [here](https://colab.research.google.com/gist/ravikyram/2ace9e7eb906a0159c27748a0b159434/untitled305.ipynb). Thanks!

**Comment by Saduf2019** [comment]
Issue is replicating with TF nightly version (2.6.0-dev20210416).Please, find the [gist here](https://colab.research.google.com/gist/Saduf2019/74c25be65ca20bc5653aac311d0038b5/untitled590.ipynb). Thanks!

**Comment by eli-osherovich** [comment]
It seems that the problem still exists in TF 2.5.

Any plans to fix it?

**Comment by sanatmpa1** [comment]
Able to reproduce the issue in `TF 2.6` and `TF-nightly`, and the reported issue still exists. Please find the [gist here](https://colab.research.google.com/gist/sanatmpa1/24747a43ef076d4941d017d44db14d02/32322.ipynb). Thanks!

**Comment by mohantym** [comment]
Hi @deasmhumhna , Edited your code at this part to resolve the issue.
```
class RandomMaskingNoAlter(tf.keras.layers.Layer):

  def call(self, inputs):
    return inputs
```

to

```
class RandomMaskingNoAlter(tf.keras.layers.Layer):

  def call(self, inputs):
    return inputs+0
```

providing [gist](https://colab.research.google.com/gist/mohantym/e746a27163286c6424fc58b93df76ae5/32322.ipynb#scrollTo=WKwaGW2QJoZu) for reference

**Comment by eli-osherovich** [comment]
This is a not-so-nice hack to somehow circumvent the bug. 
Definitely, not a solution.

**Comment by tensorflowbutler** [comment]
Hi There,

 This is a stale issue. As you are using an older version of tensorflow, we are checking to see if you still need help on this issue. Please test the issue with the latest TensorFlow (TF2.7 and tf-nightly). If the issue still persists with the newer versions of TF, please feel free to open it in [keras-team/keras](https://github.com/keras-team/keras/issues) repository by providing details about the issue and a standalone code to reproduce the issue. Thanks! 

 Please note that Keras development has moved to a separate Keras-team/keras repository to focus entirely on only Keras. Thanks!

**Comment by eli-osherovich** [comment]
I would suggest that Keras team will add a unit test to cover this issue (and endless number of other issues). Passing down to the users the task of testing issues in every new release with a hope that the issue is somehow fixed does not look like a good strategy.....

**Comment by stevetracvc** [comment]
Thanks @mohantym that saved me. I was wondering why mine wasn't working either, and that fixed it for me. I agree, this is a crappy fix though. I'm running TF 2.6.0 installed via Conda and it's an issue there, but I can't check a newer version

**Comment by tilakrayal** [comment]
Hi,

Thank you for opening this issue. Since this issue has been open for a long time, the code/debug information for this issue may not be relevant with the current state of the code base.

The Tensorflow team is constantly improving the framework by fixing bugs and adding new features. We suggest you try the latest TensorFlow [version](https://pypi.org/project/tensorflow/) with the latest compatible hardware configuration which could potentially resolve the issue. If you are still facing the issue, please create a new GitHub issue with your latest findings, with all the debugging information which could help us investigate.

Please follow the [release notes](https://github.com/tensorflow/tensorflow/blob/master/RELEASE.md) to stay up to date with the latest developments which are happening in the Tensorflow space.
